tidy3d/plugins/design/method.py

```diff
@@ -175,28 +110,466 @@ class MethodGrid(MethodIndependent):
+            description="The style of gene mutation.",
+        )
+    )
+
+    mutation_prob: Optional[pd.confloat(ge=0, le=1)] = pd.Field(
+        default=0.2,
+        title="Crossover probability.",
+        description="The probability of mutating a gene.",
+    )
+
+    # TODO: See if anyone is interested in having the full suite of PyGAD options - there's a lot!
+
+    def run(self, parameters: Tuple[ParameterType, ...], run_fn: Callable, console) -> Tuple[Any]:
+        """Defines the search algorithm for the GA"""
+        try:
+            import pygad
+        except ImportError:
+            raise ImportError(
+                "Cannot run genetic algorithm optimization as pygad module not found. Please check installation."
+            )
+
+        # Make param names available to the fitness function
+        param_keys = [param.name for param in parameters]
+
+        # Store parameters and fitness
+        store_parameters = []
+        store_fitness = []
+        store_aux = []
+
+        # Set gene_spaces to keep GA within ranges
+        param_converter = {}
+        gene_spaces = []
+        gene_types = []
+        for param in parameters:
+            if type(param) == ParameterFloat:
+                gene_spaces.append({"low": param.span[0], "high": param.span[1]})
+                gene_types.append(float)
+            elif type(param) == ParameterInt:
+                gene_spaces.append(
+                    range(param.span[0], param.span[1] + 1)
+                )  # +1 included so as to be inclusive of upper range value
+                gene_types.append(int)
+            else:
+                # Designed for str in ParameterAny but may work for anything
+                param_converter[param.name] = self.any_to_int_param(param)
+
+                gene_spaces.append(range(0, len(param.allowed_values)))
+                gene_types.append(int)
+
+        # Create fitness function combining pre and post fn with the tidy3d call
+        def fitness_function(ga_instance, solution, solution_idx):
+            # Break solution down to dict
+            sol_dict = self.sol_array_to_dict(solution, param_keys, param_converter)
+
+            # Iterate through solutions for batched and non-batched data
+            sol_out, aux_out = self._extract_output(run_fn(sol_dict))
+
+            self._flatten_and_append(aux_out, store_aux)
+
+            return sol_out
+
+        def on_generation(ga_instance):
+            store_parameters.append(ga_instance.population.copy())
+            store_fitness.append(ga_instance.last_generation_fitness)
+            if console is not None:
+                best_fitness = ga_instance.best_solution()[1]
+                console.log(
+                    f"Generation {ga_instance.generations_completed} Best Fitness: {best_fitness:.3f}"
+                )
+
+        # Determine initial array
+        num_genes = len(parameters)
+
+        # define the optimizer
+        ga_instance = pygad.GA(
```

Comment:
consider making this a `@property`. then the user could even call eg. `design_space.method.optimizer` and see this.

Reply:
I'm slightly lost as to what this would allow? Why would the user need to interact with this?

Reply:
just wondering if the user would want to inspect the GA [class](https://pygad.readthedocs.io/en/latest/pygad.html#pygad-ga-class) before or after running. For example to see how things are set up or plot results.

For example

```py
des = DesignSpace(..., method=method_ga)

print(des.optimizer)

des.run()

```

Reply:
Ah I see what you mean. They do have access to the optimizer object after the run as it's now output as Results.optimizer so people can plot and analyze it. 

I don't think we should give access to it before the optimization as it doesn't exist before method.run is called, and all the relevant parameters should be set with Fields when the method is created.

Reply:
ah ok that makes sense. 👍